workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ComposeLifecycleOwner.kt

```diff
@@ -0,0 +1,103 @@
+ */
+@Composable internal fun rememberChildLifecycleOwner(
+  parentLifecycle: Lifecycle = LocalLifecycleOwner.current.lifecycle
+): LifecycleOwner {
+  val lifecycleOwner = remember {
+    ComposeLifecycleOwner.installOn(parentLifecycle)
+  }
+  return lifecycleOwner
+}
+
+/**
+ * A custom [LifecycleOwner] that synchronizes its lifecycle with a parent [Lifecycle] and
+ * integrates with Jetpack Compose's lifecycle through [RememberObserver].
+ *
+ * ## Purpose
+ *
+ * - Ensures that any lifecycle-aware components within a composable function have a lifecycle that
+ *   accurately reflects both the parent lifecycle and the composable's own lifecycle.
+ * - Manages lifecycle transitions and observer registration/removal to prevent memory leaks and
+ *   ensure proper cleanup when the composable leaves the composition.
+ *
+ * ## Key Features
+ *
+ * - Lifecycle Synchronization: Mirrors lifecycle events from the provided `parentLifecycle` to
+ *   its own [LifecycleRegistry], ensuring consistent state transitions.
+ * - Compose Integration: Implements [RememberObserver] to align with the composable's lifecycle
+ *   in the Compose memory model.
+ * - Automatic Observer Management: Adds and removes a [LifecycleEventObserver] to the parent
+ *   lifecycle, preventing leaks and ensuring proper disposal.
+ * - **State Transition Safety:** Carefully manages lifecycle state changes to avoid illegal
+ *   transitions, especially during destruction.
+ *
+ * ## Usage Notes
+ *
+ * - Should be used in conjunction with `remember` and provided the `parentLifecycle` as a key to
+ *   ensure it updates correctly when the parent lifecycle changes.
+ * - By integrating with Compose's lifecycle, it ensures that resources are properly released when
+ *   the composable leaves the composition.
+ *
+ * @param parentLifecycle The parent [Lifecycle] with which this lifecycle owner should synchronize.
+ */
+private class ComposeLifecycleOwner(
+  private val parentLifecycle: Lifecycle
+) : LifecycleOwner, RememberObserver, LifecycleEventObserver {
+
+  private val registry = LifecycleRegistry(this)
+  override val lifecycle: Lifecycle
+    get() = registry
+
+  override fun onRemembered() {
+  }
+
+  override fun onAbandoned() {
+    onForgotten()
+  }
+
+  override fun onForgotten() {
+    parentLifecycle.removeObserver(this)
+
+    // If we're leaving the composition, ensure the lifecycle is cleaned up
+    if (registry.currentState != Lifecycle.State.INITIALIZED) {
+      registry.currentState = Lifecycle.State.DESTROYED
+    }
+  }
+
+  override fun onStateChanged(
+    source: LifecycleOwner,
+    event: Event
+  ) {
+    registry.handleLifecycleEvent(event)
+  }
+
+  companion object {
+    fun installOn(parentLifecycle: Lifecycle): ComposeLifecycleOwner {
+      return ComposeLifecycleOwner(parentLifecycle).also {
+        // We need to synchronize the lifecycles before the child ever even sees the lifecycle
```

Comment:
i want to inline this comment appropriately deferring against use method reference for this one